File: PongSharp/src/Models/Platform.cs
using System.Numerics;
using PongSharp.Data;

namespace PongSharp.Models;

internal class Platform
{
    private readonly CollisionHandler _collisionHandler;
    private Vector2 _position;
    private readonly Vector2 _mapSize;
    private readonly Vector2 _size;
    private readonly float _velocity;

    public Platform(CollisionHandler collisionHandler, Vector2 position, Vector2 mapSize)
    {
        _collisionHandler = collisionHandler;
        _position = position;
        _mapSize = mapSize;
        _size = new Vector2(1, 7);
        _velocity = 20;
        Rectangle = new Rectangle(_position, _size);
    }

    public Rectangle Rectangle { get; private set; }

    public Vector2 Direction { get; private set; }

    public void Move(float deltaTime, Vector2 direction)
    {
        var newPosition = _position + direction * deltaTime * _velocity;
        var clampedPosition = Vector2.Clamp(newPosition, Vector2.Zero, _mapSize - _size);
        var newRectangle = new Rectangle(clampedPosition, _size);

        _position = clampedPosition;
        Rectangle = newRectangle;
        Direction = direction;
    }

    public void ResetDirection() => Direction = Vector2.Zero;
}

File: Pong/src/Models/Platform.c
#include <stdlib.h>
#include "Float2.h"
#include "RectangleF.h"
#include "Platform.h"
#include "Int2.h"
#include "CollisionHandler.h"
#include "GameData.h"

Platform* Platform_Create(Float2 position)
{
    Platform* result = calloc(1, sizeof(Platform));
    Float2 size = {1, 7};
    RectangleF rect = {position, size};

    result->_position = position;
    result->_size = size;
    result->_velocity = 20;
    result->Rectangle = rect;

    return result;
}

void Platform_Dispose(Platform* platform)
{
    free(platform);
}

void Platform_Move(Platform* platform, float deltaTime, Float2 direction)
{
    Float2 newPosition = Float2_Add(platform->_position, Float2_MulFloat(direction, deltaTime * platform->_velocity));
    Float2 clampedPosition = Float2_Clamp(newPosition, Float2_Zero(), Float2_Sub(GameStaticData.MapSize, platform->_size));
    RectangleF newRectangle = {clampedPosition, platform->_size};

    platform->_position = clampedPosition;
    platform->Rectangle = newRectangle;
    platform->Direction = direction;
}

void Platform_ResetDirection(Platform* platform)
{
    platform->Direction = Float2_Zero();
}